### PR TITLE
C#: Improve the `unsafe` predicate on Modifiable.

### PR DESCRIPTION
**Pull request by michaelnebel**
It turns out that the `unsafe` keyword is discarded during compilation, which means that when extracting from referenced assemblies, we are not able to identify whether a member is `unsafe`.
The compiler requires methods, delegates and properties that make any use of pointer like types is declared unsafe.
As this might not be that relevant for our queries it is relevant for stub generation.

One example of a *nasty* delegate type declared in .NET Runtime (ie. a `delegate` with a `delagate*` (function pointer) return type).
```csharp
// Generated from `System.Runtime.InteropServices.ObjectiveC.ObjectiveCMarshal+UnhandledExceptionPropagationHandler` in `System.Runtime.InteropServices, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a`
unsafe public delegate delegate* unmanaged<System.IntPtr, void> UnhandledExceptionPropagationHandler(System.Exception exception, System.RuntimeMethodHandle lastMethod, out System.IntPtr context);
```

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/Member.qll [hunk 1]
```diff
@@ -98,10 +98,21 @@ class Modifiable extends Declaration, @modifiable {
...
-    this.(Callable).getReturnType() instanceof PointerType
+    this.hasModifier("unsafe")
+    or
+    exists(Type t, Type child |
+      t = this.(Parameterizable).getAParameter().getType() or
+      t = this.(Property).getType() or
+      t = this.(Callable).getReturnType() or
+      t = this.(DelegateType).getReturnType()
+    |
+      child = t.getAChild*() and
```
Comment (tamasvajk) [inline csharp/ql/lib/semmle/code/csharp/Member.qll:109]: Why is the `getAChild` needed?

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/Member.qll:109]: I think the example on the PR description illustrates why it is needed.

Comment (tamasvajk) [reply csharp/ql/lib/semmle/code/csharp/Member.qll:109]: Isn't that just a `delegate` whose return type is a function pointer? Wouldn't `this.(DelegateType).getReturnType() instanceof FunctionPointerType` hold there?

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/Member.qll:109]: For now the `getAChild` only has an effect on the actual .NET 7 stubs being created for arrays of pointers.
An example could be
```csharp
public void FromManaged(T*[] array, System.Span<TUnmanagedElement> buffer)
```

File: csharp/ql/test/query-tests/Stubs/All/Test.cs [hunk 2]
```diff
@@ -172,6 +172,8 @@ public class Class11 : IInterface2<Class11>, IInterface3<Class11>
         static explicit IInterface2<Class11>.operator int(Class11 n) => 0;
     }
 
+    public static unsafe class MyUnsafeClass { }
```
Comment (hvitved) [inline csharp/ql/test/query-tests/Stubs/All/Test.cs:175]: Should we have a test for something like
```
public void FromManaged(T*[] array, System.Span<TUnmanagedElement> buffer)
```
as well?

**Review by hvitved — APPROVED** [review]

**Comment by michaelnebel** [comment]
DCA looks good.

**Pull request merged**